### PR TITLE
Add support for RetryPolicy in test framework

### PR DESCRIPTION
**Pull request by yiminc**
Add support for RetryPolicy in test framework.

### CODE REVIEW

File: internal/internal_workflow_testsuite.go [hunk 1]
```diff
@@ -684,6 +697,35 @@ func (env *testWorkflowEnvironmentImpl) Complete(result []byte, err error) {
...
+		var expireTime time.Time
+		if params.retryPolicy.GetExpirationIntervalInSeconds() > 0 {
+			expireTime = params.scheduledTime.Add(time.Second * time.Duration(params.retryPolicy.GetExpirationIntervalInSeconds()))
+		}
+		backoff := getRetryBackoffFromThriftRetryPolicy(params.retryPolicy, env.workflowInfo.Attempt, errReason, env.Now(), expireTime)
+		if backoff > 0 {
+			env.registerDelayedCallback(func() {
+				// remove the current child workflow from the pending child workflow map because
+				// the childWorkflowID will be the same for retry run.
+				delete(env.runningWorkflows, env.workflowInfo.WorkflowExecution.ID)
```
Comment (mfateev) [inline internal/internal_workflow_testsuite.go]: What's going to happen if signal is sent to a child while waiting for a retry? I think in real life it should be delivered to the next run.

File: internal/internal_workflow_testsuite_test.go [hunk 2]
```diff
@@ -1792,3 +1792,137 @@ func (s *WorkflowTestSuiteUnitTest) Test_ContextMisuse() {
...
+}
+
+func (s *WorkflowTestSuiteUnitTest) Test_LocalActivityRetry() {
+
+	localActivityFn := func(ctx context.Context) (string, error) {
+		info := GetActivityInfo(ctx)
+		if info.Attempt < 2 {
+			return "", NewCustomError("bad-luck")
+		}
+		return "retry-done", nil
```
Comment (mfateev) [inline internal/internal_workflow_testsuite_test.go]: May be return the non retryable error? Or have a test just for it?

File: internal/internal_workflow_testsuite.go (hunk 1 above)
Comment (yiminc) [reply internal/internal_workflow_testsuite.go]: Just realize that currently, there is no way to send signal to child workflow in the unit test environment. https://github.com/uber-go/cadence-client/issues/564 will address that separately.

File: internal/internal_workflow_testsuite_test.go (hunk 2 above)
Comment (yiminc) [reply internal/internal_workflow_testsuite_test.go]: will update to cover non-retryable error case.

File: internal/internal_workflow_testsuite_test.go [hunk 3]
```diff
@@ -1808,18 +1816,23 @@ func (s *WorkflowTestSuiteUnitTest) Test_ActivityRetry() {
...
 				BackoffCoefficient:       2,
-				NonRetriableErrorReasons: []string{"bad-error"},
+				NonRetriableErrorReasons: []string{"bad-bug"},
 				ExpirationInterval:       time.Minute,
 			},
 		}
 		ctx = WithActivityOptions(ctx, ao)
 
+		err := ExecuteActivity(ctx, activityFailedFn).Get(ctx, nil)
+		if err == nil {
```
Comment (mfateev) [inline internal/internal_workflow_testsuite_test.go]: May be add the check that it is indeed "bad-bug"

**Pull request merged**